File: src/Components/HmacHeaderBuilder.cs
using HmacManager.Exceptions;
using HmacManager.Extensions;
using HmacManager.Headers;
using HmacManager.Mvc;

namespace HmacManager.Components;

/// <summary>
/// A class representing a builder to create required hmac headers to be added to a request.
/// </summary> 
public class HmacHeaderBuilder : IHmacHeaderBuilder
{
    /// <summary>
    /// A dictionary of <see cref="HmacAuthenticationDefaults.Headers"/> defaulting to null.
    /// </summary> 
    protected readonly IDictionary<string, string?> HeaderValues = new Dictionary<string, string?>
    {
        { HmacAuthenticationDefaults.Headers.Authorization, null },
        { HmacAuthenticationDefaults.Headers.Policy, null },
        { HmacAuthenticationDefaults.Headers.Scheme, null },
        { HmacAuthenticationDefaults.Headers.Nonce, null },
        { HmacAuthenticationDefaults.Headers.DateRequested, null }
    };

    /// <summary>
    /// An enumerable of header values that are non empty, i.e. both keys and values are not null, empty or whitespace.
    /// </summary>
    protected IEnumerable<KeyValuePair<string, string?>> NonEmptyHeaderValues => HeaderValues.Where(element => element.NonEmpty());

    /// <summary>
    /// Creates an instance of <c>HmacHeaderBuilder</c>.
    /// </summary>
    public HmacHeaderBuilder()
    {
    }

    /// <summary>
    /// Creates an instance of <c>HmacHeaderBuilder</c>.
    /// </summary>
    /// <param name="options">An instance of <c>HmacManagerOptions</c>.</param>
    /// <param name="hmac">An instance of <c>Hmac</c>.</param>
    public HmacHeaderBuilder(HmacManagerOptions options, Hmac hmac)
    {
        WithPolicy(options.Policy);
        
        if (!string.IsNullOrWhiteSpace(options.Scheme?.Name))
        {
            WithScheme(options.Scheme.Name);
        }

        WithNonce(hmac.Nonce);
        WithDateRequested(hmac.DateRequested);

        if (string.IsNullOrWhiteSpace(hmac.Signature))
        {
            throw new Exception();
        }

        WithAuthorization(hmac.Signature);
    }

    /// <summary>
    /// Adds the specified signature to the builder.
    /// </summary>
    /// <param name="signature">A string representing the signature.</param>
    /// <returns>The builder.</returns>    
    public IHmacHeaderBuilder WithAuthorization(string signature)
    {
        HeaderValues[HmacAuthenticationDefaults.Headers.Authorization] = $"{HmacAuthenticationDefaults.AuthenticationScheme} {signature}";
        return this;
    }

    /// <summary>
    /// Adds the specified policy to the builder.
    /// </summary>
    /// <param name="policy">A string representing the policy.</param>
    /// <returns>The builder.</returns>    
    public IHmacHeaderBuilder WithPolicy(string policy)
    {
        HeaderValues[HmacAuthenticationDefaults.Headers.Policy] = policy;
        return this;
    }
    
    /// <summary>
    /// Adds the specified scheme to the builder.
    /// </summary>
    /// <param name="scheme">A string representing the scheme.</param>
    /// <returns>The builder.</returns> 
    public IHmacHeaderBuilder WithScheme(string scheme)
    {
        HeaderValues[HmacAuthenticationDefaults.Headers.Scheme] = scheme;
        return this;
    }

    /// <summary>
    /// Adds the specified nonce to the builder.
    /// </summary>
    /// <param name="nonce">The nonce represented by a guid.</param>
    /// <returns>The builder.</returns> 
    public IHmacHeaderBuilder WithNonce(Guid nonce)
    {
        HeaderValues[HmacAuthenticationDefaults.Headers.Nonce] = nonce.ToString();
        return this;
    }

    /// <summary>
    /// Adds the specified date and time to the builder.
    /// </summary>
    /// <param name="dateRequested">The date requested.</param>
    /// <returns>The builder.</returns>
    public IHmacHeaderBuilder WithDateRequested(DateTimeOffset dateRequested)
    {
        HeaderValues[HmacAuthenticationDefaults.Headers.DateRequested] = dateRequested.ToUnixTimeMilliseconds().ToString();
        return this;
    }

    /// <summary>
    /// Creates a new instance of <c>HmacHeaderBuilder</c>.
    /// </summary>
    /// <returns>An instance of <c>HmacHeaderBuilder</c>.</returns>
    public virtual IHmacHeaderBuilder CreateBuilder(HmacManagerOptions options, Hmac hmac) => new HmacHeaderBuilder(options, hmac);

    /// <summary>
    /// Builds the collection of header values.
    /// </summary>
    /// <returns>A readonly collection of header values.</returns>
    public virtual IReadOnlyCollection<HeaderValue> Build()
    {
        if (NonEmptyHeaderValues.Any())
        {
            return NonEmptyHeaderValues.Select(header => new HeaderValue(header)).ToList();
        }
        else
        {
            throw new MissingHeaderException();
        }
    }
}

File: client/lib/hmac_manager/builders/hmac-header-builder.ts
import { HmacAuthenticationDefaults } from "../hmac-authentication-defaults.js";
import { HmacHeaderCollection } from "./hmac-options-header-builder.js";

/**
 * A builder class to facilitate the construction of HMAC authentication headers.
 * This class allows for the fluent construction of HMAC headers, with each header 
 * being set individually before building the final header collection.
 * 
 * Example usage:
 * ```typescript
 * const builder = new HmacHeaderBuilder();
 * const headers = builder
 *     .withAuthorization('signature123')
 *     .withPolicy('policyABC')
 *     .withScheme('HMAC')
 *     .withNonce('nonce123')
 *     .withDateRequested('2024-11-15')
 *     .build();
 * ```
 */
export default class HmacHeaderBuilder {
    /**
     * The collection of header values used in HMAC authentication.
     * These values are initially set to `null` and can be updated via the
     * provided methods.
     */
    protected readonly HeaderValues: HmacHeaderCollection = {
        [HmacAuthenticationDefaults.Headers.Authorization]: null,
        [HmacAuthenticationDefaults.Headers.Policy]: null,
        [HmacAuthenticationDefaults.Headers.Scheme]: null,
        [HmacAuthenticationDefaults.Headers.Nonce]: null,
        [HmacAuthenticationDefaults.Headers.DateRequested]: null
    };

    /**
     * Retrieves the header values that are not null.
     * This method filters out any headers that have `null` as their value.
     * 
     * @returns A new object containing only the header values that are set.
     */
    protected get NonEmptyHeaderValues(): HmacHeaderCollection {
        const nonEmptyHeaderValues: HmacHeaderCollection = {};
        for (const [key, value] of Object.entries(this.HeaderValues)) {
            if (value !== null) {
                nonEmptyHeaderValues[key as HmacAuthenticationDefaults.Headers] = value;
            }
        }
        return nonEmptyHeaderValues;
    }

    /**
     * Sets the `Authorization` header with the provided signature.
     * The signature is prefixed with the default authentication scheme.
     * 
     * @param signature The HMAC signature to be included in the `Authorization` header.
     * @returns The builder instance for fluent chaining.
     */
    withAuthorization(signature: string): HmacHeaderBuilder {
        this.HeaderValues[HmacAuthenticationDefaults.Headers.Authorization] = `${HmacAuthenticationDefaults.AuthenticationScheme} ${signature}`;
        return this;
    }

    /**
     * Sets the `Policy` header with the provided policy value.
     * 
     * @param policy The policy to be included in the `Policy` header.
     * @returns The builder instance for fluent chaining.
     */
    withPolicy(policy: string): HmacHeaderBuilder {
        this.HeaderValues[HmacAuthenticationDefaults.Headers.Policy] = policy;
        return this;
    }

    /**
     * Sets the `Scheme` header with the provided scheme value.
     * 
     * @param scheme The scheme to be included in the `Scheme` header.
     * @returns The builder instance for fluent chaining.
     */
    withScheme(scheme: string | null): HmacHeaderBuilder {
        this.HeaderValues[HmacAuthenticationDefaults.Headers.Scheme] = scheme;
        return this;
    }

    /**
     * Sets the `Nonce` header with the provided nonce value.
     * 
     * @param nonce The nonce to be included in the `Nonce` header.
     * @returns The builder instance for fluent chaining.
     */
    withNonce(nonce: string): HmacHeaderBuilder {
        this.HeaderValues[HmacAuthenticationDefaults.Headers.Nonce] = nonce;
        return this;
    }

    /**
     * Sets the `DateRequested` header with the provided date value.
     * 
     * @param dateRequested The date requested to be included in the `DateRequested` header.
     * @returns The builder instance for fluent chaining.
     */
    withDateRequested(dateRequested: Date): HmacHeaderBuilder {
        this.HeaderValues[HmacAuthenticationDefaults.Headers.DateRequested] = dateRequested.getTime().toString();
        return this;
    }

    /**
     * Builds and returns the final header collection, excluding any headers with `null` values.
     * 
     * @returns A collection of headers with only non-null values.
     */
    build = (): HmacHeaderCollection => this.NonEmptyHeaderValues;
}
